src/Rings/mpoly-graded.jl

```diff
@@ -2012,38 +2012,33 @@ function _hilbert_numerator_bayer_stillman(
   #
   # Algorithm 2.6, page 35
   ###########################################################################
-  ret = _hilbert_numerator_trivial_cases(a, weight_matrix, S)
+  ret = _hilbert_numerator_trivial_cases(a, weight_matrix, S, t, oneS)
   ret !== nothing && return ret
 
-  t = gens(S)
-
   # make sure we have lexicographically ordered monomials
-  sort!(a)
+  sort!(a, alg=QuickSort)
 
   # initialize the result 
-  h = one(S) - prod(t[i]^sum([a[1][j]*weight_matrix[i, j] for j in 1:length(a[1])]) for i in 1:length(t))
+  h = oneS - _expvec_to_poly(S, t, weight_matrix, a[1])
+  linear_mons = Vector{Int}()
   for i in 2:length(a)
     J = _divide_by(a[1:i-1], a[i])
-    J1 = Vector{Vector{Int}}()
-    linear_mons = Vector{Int}()
-    for m in J
-      if sum(m) > 1
-        push!(J1, m)
-      else
-        for k in 1:length(m)
-          if m[k] > 0 
-            push!(linear_mons, k)
-            break
-          end
-        end
+    empty!(linear_mons)
+    J1 = filter!(J) do m
+      k = findfirst(!iszero, m)
+      k === nothing && return false # filter out zero vector
+      if m[k] == 1 && findnext(!iszero, m, k + 1) === nothing
+        push!(linear_mons, k)
+        return false
       end
+      return true
     end
-    q = _hilbert_numerator_bayer_stillman(J1, weight_matrix, S)
+    q = _hilbert_numerator_bayer_stillman(J1, weight_matrix, S, t, oneS)
     for k in linear_mons
-      q = q*(one(S) - prod(t[i]^weight_matrix[i, k] for i in 1:length(t)))
+      q *= (oneS - prod(t[i]^weight_matrix[i, k] for i in 1:length(t)))
```

Comment:
I found that using `*=` was breaking my code in various other occasions. One really has to be careful not to modify polynomials which are then also used elsewhere. Plus: As far as I remember, there was no speedup for multivariate polynomials for this. 

But here, the usage was OK?

Reply:
Quoting the documentation:
> `x *= y` is a synonym for `x = x * y`

So this should not break anything (unless there's a bug).

It's just more concise and IMHO makes it a bit more readable (i.e. with `q *= SOMETHING` it is immediately clear that `q` is multiplied by `SOMETHING`; but with `q = q * SOMETHING` I have to read the whole line carefully, as it could also be `q = q * STUFF + MORE`